functions/src/aqi-calculation/purple-air-response.ts

```diff
@@ -1,32 +1,370 @@
+        longitude: longitude,
+        meanPercentDifference: meanPercentDifference,
+        pm25: pm25,
+        humidity: humidity,
+        timestamp: timestamp,
+      },
+    ];
+  } else {
+    return [id, null];
   }
 }
+
+/**
+ * Translates PurpleAir response into map of sensor ID to PurpleAirReading
+ * @returns map of sensor ID to PurpleAirReading
+ */
+async function getReadingsMap(): Promise<Map<number, PurpleAirReading | null>> {
+  const purpleAirResponse = await fetchPurpleAirResponse();
+  const readings: Map<number, PurpleAirReading | null> = new Map();
+  const purpleAirData = purpleAirResponse.data;
+  const fieldNames: string[] = purpleAirData.fields;
+  const rawReadings: (string | number)[][] = purpleAirData.data;
+  rawReadings.forEach(rawReading => {
+    const [id, reading] = getReading(rawReading, fieldNames);
+    readings.set(id, reading);
+  });
+  return readings;
+}
+
+/**
+ * For a given readings subcollection, gets the most recent reading Timestamp
+ * @param readingsCollectionRef - reference to readings collection for sensor to get the most recent reading time
+ * @returns a Promise of the Timestamp of the most recent sensor reading time, or null if no readings
+ */
+async function getLastSensorReadingTime(
+  readingsCollectionRef: FirebaseFirestore.CollectionReference<FirebaseFirestore.DocumentData>
+): Promise<FirebaseFirestore.Timestamp | null> {
+  let lastSensorReadingTime: FirebaseFirestore.Timestamp | null = null;
+  const maxDocs = 1;
+  const querySnapshot = await readingsCollectionRef
+    .orderBy('timestamp', 'desc')
+    .limit(maxDocs)
+    .get();
+  // There should only be one document in docs, but loops over docs since it's an array
+  for (const sensorDoc of querySnapshot.docs) {
+    if (sensorDoc.data().timestamp) {
+      lastSensorReadingTime = sensorDoc.data().timestamp;
+    }
+  }
+  return lastSensorReadingTime;
+}
+
+/**
+ * Converts the PurpleAir ID to a number, if necessary.
+ * @param id - PurpleAir ID stored as a string or number
+ * @returns the PurpleAir ID as a number
+ */
+function getPurpleAirId(id: string | number): number {
+  if (typeof id === 'string') return +id;
+  return id;
+}
+
+/**
+ * Handles the PurpleAir API call for all active sensors,
+ * storing the new data in each sensors' readings collection.
+ */
+async function purpleAirToFirestore(): Promise<void> {
+  const readingsMap = await getReadingsMap();
+
+  // Add each of the new readings to the readings subcollection and the pm25buffers
+  const activeSensorDocsSnapshot = await firestore
+    .collection('sensors')
+    .where('isActive', '==', true)
+    .get();
+
+  for (const sensorDoc of activeSensorDocsSnapshot.docs) {
+    const readingsCollectionRef = firestore.collection(
+      readingsSubcollection(sensorDoc.id)
+    );
+
+    // Initialize the buffer element to the default value
+    let pm25BufferElement: Pm25BufferElement = defaultPm25BufferElement;
+
+    // Get the reading for this sensor
```

Comment:
Not getting "the reading"? This is getting all of a sensor's current data. Change to something like "Get the data for this sensor"